Dockerfile

```diff
@@ -54,6 +54,10 @@ RUN \
   postfix-policyd-spf-python postsrsd pyzor \
   razor rpm2cpio rsyslog sasl2-bin spamassassin supervisor \
   unrar-free unzip whois xz-utils >/dev/null && \
+  # Fail2Ban
+  curl -Lso fail2ban.deb https://github.com/fail2ban/fail2ban/releases/download/0.11.2/fail2ban_0.11.2-1.upstream1_all.deb && \
```

Comment:
If we do this to be on forefront of security then we should check the downloaded files integrity/signing shouldn't we?

Reply:
Yes, we should do that. Can you help?

Reply:
Give me a minute

Reply:
One easy method, without using GPG, would be to check the pre-computed SHA152 sum like this:

``` BASH
sha512sum --check --status <<< "09d91ad0f499b0825b9d6da1d175400aa32498aff5705624e88b3d7e02cf727403c49b967c732e42182e25d4f439de0faf681f9f6d3598364e3f2f2e3453f8bf  fail2ban_0.11.2-1.upstream1_all.deb"
``` 

which could be written directly into the Dockerfile.

Reply:
This should be enough, to ensure the file was not altered.

However, my checksum differs:

`echo "44fe3d67e7d24bba4fc81ee672059dcb67980258db6101bcd8740736a9d9110f *fail2ban.deb" | sha256sum --check --status`

or

`sha256sum --check --status <<< "44fe3d67e7d24bba4fc81ee672059dcb67980258db6101bcd8740736a9d9110f *fail2ban.deb"`

Reply:
Here is an (truncated Dockerfile) example using pgp:

```
ARG FAIL2BAN_VERSION=0.11.2
ARG FAIL2BAN_URL=https://github.com/fail2ban/fail2ban/releases/download/${FAIL2BAN_VERSION}/fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb
ARG FAIL2BAN_URL_ASC=https://github.com/fail2ban/fail2ban/releases/download/${FAIL2BAN_VERSION}/fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb.asc

RUN apt-get install -q -y --no-install-recommends gpg=2.2.12-1+deb10u1 \
 && wget ${FAIL2BAN_URL} \
 && wget ${FAIL2BAN_URL_ASC} \
 && gpg --verify fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb.asc fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb
 && dpkg -i fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb
 # && apt cleanup of gpg etc
```

If someone can provide me the fail2ban public key and the fingerprint we could additionally check the fingerprint (like i did [here](https://github.com/wernerfred/docker-rainloop/blob/755a7776bb9fbc6bcdc427acc1d303a42dd84956/Dockerfile#L33))

Reply:
@casperklein You used `sha256sum`, I used `sha512sum`.

@wernerfred Your solution looks very good to me. As you have made yourself the work of writing this up, we should use it (and of course, since it is better than just checking a hash sum). You're right, we need to get the pubkey and fingerprint. Shall I open an issue in the F2B repo and ask?

Reply:
I did a short research but didn't find it directly. Maybe it is the easiest approach to ask directly

Reply:
@wernerfred Have a look at the issue that is linked to the F2B repo. A contributor explained how to get pubkey and fingerprint.

Reply:
This should be it:
```
ARG FAIL2BAN_VERSION=0.11.2
ARG FAIL2BAN_URL=https://github.com/fail2ban/fail2ban/releases/download/${FAIL2BAN_VERSION}/fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb
ARG FAIL2BAN_URL_ASC=https://github.com/fail2ban/fail2ban/releases/download/${FAIL2BAN_VERSION}/fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb.asc
ARG FAIL2BAN_PGP_PUBLIC_KEY_ID=0x683BF1BEBD0A882C
ARG FAIL2BAN_PGP_PUBLIC_KEY_SERVER=keys.gnupg.net
ARG FAIL2BAN_GPG_FINGERPRINT="8738 559E 26F6 71DF 9E2C 6D9E 683B F1BE BD0A 882C"

RUN apt-get install -q -y --no-install-recommends wget=1.20.1-1.1 gpg=2.2.12-1+deb10u1 gpg-agent=2.2.12-1+deb10u1\
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/* \
 && gpg --keyserver ${FAIL2BAN_PGP_PUBLIC_KEY_SERVER} --recv-keys ${FAIL2BAN_PGP_PUBLIC_KEY_ID}
 && wget -q ${FAIL2BAN_URL} \
 && wget -q ${FAIL2BAN_URL_ASC} \
 && FINGERPRINT="$(LANG=C gpg --verify fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb.asc fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb 2>&1 \
    | sed -n "s#Primary key fingerprint: \(.*\)#\1#p")" \
 && if [ -z "${FINGERPRINT}" ]; then echo "ERROR: Invalid GPG signature!" && exit 1; fi \
 && if [ "${FINGERPRINT}" != "${FAIL2BAN_GPG_FINGERPRINT}" ]; then echo "ERROR: Wrong GPG fingerprint!" && exit 1; fi \
 && dpkg -i fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb
 && rm -rf fail2ban_${FAIL2BAN_VERSION}-1.upstream1_all.deb
 && apt-get purge -q -y wget gpg gpg-agent \
 && apt-get autoremove -y \
```

I have no time to test right now, might be that we need a bit more tweaking on cleanup or so. Also the `${FAIL2BAN_VERSION}` should be reworked bc other relases of F2B do not have the `-1.upstream` which is hard coded in my example.

I am able to test this today evening/night. If someone wants to do this earlier: do it 👍🏻

Reply:
I have prepared the Dockerfile for you, and with latest commit it should suffice. You can give it a try locally, and check if you think everything is okay. I had to change some things here and there, but the overall construction stayed the same.

Reply:
- Is there a reason for the package pinning in that particular case (we don't do this for the other packages)? If we do it that way, we are probably going to install old versions of these packages maybe in some weeks, if some does not constantly check for new versions and adjust the Dockerfile.

- `ARG` should be `ENV`

- `ENV FAIL2BAN_URL_ASC=${FAIL2BAN_URL}.asc`

> ${FAIL2BAN_VERSION} [..] which is hard coded in my example.

IMO we don't need `FAIL2BAN_VERSION` at all. There are not that much releases of f2b. And as you said, when there is a new version, we have to adjust the Dockerfile and maybe f2b_url either way.

Reply:
In context of security, this feels wrong: `wget -q --no-check-certificate`.

Serious question: What kind of improved security do we get from this, compared to the very simple file hash comparison @aendeavor suggested? With both methods, we can assure, that the .deb file was not altered.

Reply:
> In context of security, this feels wrong: `wget -q --no-check-certificate`.
> 
> 
> 
> Serious question: What kind of improved security do we get from this, compared to the very simple file hash comparison @aendeavor suggested? With both methods, we can assure, that the .deb file was not altered.

This was something I asked myself too. Maybe the hash sum is the best solution as it is simple.

The `--no-check-certificate` is needed in CI as wget behaves strangely in pipelines. I have seen this behaviour in GitLab CI too.

At the moment, I think that the hash may be the better solution in the end...

Reply:
> IMO we don't need FAIL2BAN_VERSION at all. There are not that much releases of f2b. And as you said, when there is a new version, we have to adjust the Dockerfile and maybe f2b_url either way.

Agree 👍🏻 

> In context of security, this feels wrong: wget -q --no-check-certificate. 

True but doesn't matter if the signing is correct

> What kind of improved security do we get from this, compared to the very simple file hash compariso

We are not only checking if the file got altered but also if it was created by fail2ban or did i get it wrong?
Nevertheless the hash comparison adds that bit of security compared to having no check at all that i am fine with both ☑️

Reply:
docker build passes without error or warning, i'm running the test suite right now

Reply:
> docker build passes without error or warning, i'm running the test suite right now
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/20406381/118016610-e590db80-b355-11eb-8b28-8740f3f0d40c.png)
> 
> 

I will commit some of the suggested changes, but you'd need to do final adjustments to this yourself:)

Reply:
> We are not only checking if the file got altered but also if it was created by fail2ban or did i get it wrong?

Correct. I considered this only as a temporary addition (until Debian 11 is released). So manually sign checking once and then hard-coding the checksum would be enough (way less additions to Dockerfile needed).

If we are going this route long-term, then the gpg solution is more complete and convenient 👍

> We only need them during build an therefore ARG should be enough.

Good point. I had the wrong assumption, that using ARG in Dockerfile and not providing a build-arg for that during the build would throw a warning. That is not the case.
Maybe we should think about replacing most ENV with ARG in the whole Dockerfile. The ENVs are populated by the start-mailserver.sh script. So we don't need to bake them into the image.

Reply:
> The ENVs are populated by the start-mailserver.sh script. So we don't need to bake them into the image.

Good point